Code/GraphMol/Descriptors/EEM.cpp

```diff
@@ -0,0 +1,237 @@
+    for(long int i = 0; i < n; i++) {
+        double t = EEMatoms.EEMatomtype[i];
+        int idx = EEMatoms.Atomindex[i];
+        double v;
+        if (t== 1.0) {
+                v=  B1[idx];
+        }
+        if (t== 2.0) {
+                v = B2[idx];
+        }
+        if (t== 3.0) {
+                v = B3[idx];
+        }
+
+        EEM[IDX(i, i)] = v;
+        for(long int j = i + 1; j < n; j++) {
+                EEM[IDX(i, j)] = kappa / dist3D[i*n+j];
+                EEM[IDX(j, i)] = EEM[IDX(i, j)];
+            }
+        }
+    /* Fill last column & row */
+    for(long int i = 0; i < n; i++) {
+        EEM[IDX(n, i)] = 1.0; // column
+        EEM[IDX(i, n)] = -1.0; // row
+
+    }
+    #undef IDX
+    return EEM;
+}
+
+double* getBVector(RDKit::ROMol mol, int n, EEM_arrays EEMatoms) {
+       /* Fill vector b i.e. -A */
+        double * b = new double[n+1];
+        for(int j = 0; j < n; j++) {
+          double t = EEMatoms.EEMatomtype[j];
+          int idx = EEMatoms.Atomindex[j];
+            if ( t== 1.0) {
+                b[j] = -A1[idx];
+            }
+            if (t== 2.0) {
+                b[j] = -A2[idx];
+            }
+            if (t== 3.0) {
+                b[j] = -A3[idx];
+            }
+        }
+
+        b[n] = RDKit::MolOps::getFormalCharge(mol); // sum of charges
+
+    return b;
+}
+
+
+EEM_arrays calculate_EEM_parameters(RDKit::ROMol mol, int n) {
+       /* Fill vector b i.e. -A */
+        int * a = new int[n];
+        double * b = new double[n];
+
+        for(int j = 0; j < n; j++) {
+            b[j]=getAtomtype(mol,mol.getAtomWithIdx(j));
+            a[j]=mol.getAtomWithIdx(j)->getAtomicNum();
+        }
+
+    return EEM_arrays {a,b};
+}
+
+
+/* Calculate charges for a particular kappa_data structure */
+std::vector<double> calculate_charges(RDKit::ROMol mol, double * dist3D, int numAtoms, EEM_arrays EEMatoms) {
+        double * A = new double[(numAtoms+1)*(numAtoms+1)];
+        A = getEEMMatrix(mol,dist3D,numAtoms,EEMatoms);
+        double * b = new double[numAtoms+1];
+        b = getBVector(mol, numAtoms,EEMatoms);
+
+        MatrixXd AM = Map<MatrixXd>(A, numAtoms+1, numAtoms+1);
+        VectorXd bv = Map<VectorXd>(b, numAtoms+1);
+        VectorXd Res(numAtoms+1);
+
+        FullPivLU<MatrixXd> lu(AM);
+        Res = lu.solve(bv);
+
+        std::vector<double> charges(Res.data(), Res.data() + Res.size());
+
+        for(long int aix = 0; aix < numAtoms; aix++) {
+            mol.getAtomWithIdx(aix)->setProp("_EEMCharge", charges[aix], true);
+        }
+
+       delete[] A;
+       delete[] b;
+
+    return charges;
+}
+
+
+void getEEMs(const ROMol &mol, std::vector<double> &result, int numAtoms, int confId) {
+  // 3D distance matrix
+  double * dist3D = MolOps::get3DDistanceMat(mol, confId, false, true);
+  EEM_arrays EEMatoms = calculate_EEM_parameters(mol,  numAtoms);
+  std::vector<double> charges = calculate_charges(mol, dist3D, numAtoms, EEMatoms);
```

Comment:
If you make the suggested change above where `calculate_charges()` takes the result array as an argument, you can just do:
```
result.clear();
result.resize(numAtoms);
calculate_charges(mol, dist3D, numAtoms, EEMatoms, result);
```
and skip the rest of the code that's in this function.